src/apps/wall/scanner.lua

```diff
@@ -0,0 +1,299 @@
+      eth_type = function (self) return ETH_TYPE_IPv4 end;
+      lo_addr_ptr = function (self)
+         return ffi.cast(uint8_ptr, self) + flow_key_ipv4_lo_addr_offset
+      end;
+      hi_addr_ptr = function (self)
+         return ffi.cast(uint8_ptr, self) + flow_key_ipv4_hi_addr_offset
+      end;
+   }
+})
+
+local flow_key_ipv6_size = ffi.sizeof("struct swall_flow_key_ipv6")
+assert(flow_key_ipv6_size % 4 == 0)
+
+local flow_key_ipv6_lo_addr_offset =
+   ffi.offsetof("struct swall_flow_key_ipv6", "lo_addr")
+local flow_key_ipv6_hi_addr_offset =
+   ffi.offsetof("struct swall_flow_key_ipv6", "hi_addr")
+
+local flow_key_ipv6 = ffi.metatype("struct swall_flow_key_ipv6", {
+   __index = {
+      hash = make_cdata_hash_function(flow_key_ipv6_size);
+      eth_type = function (self) return ETH_TYPE_IPv6 end;
+      lo_addr_ptr = function (self)
+         return ffi.cast(uint8_ptr, self) + flow_key_ipv6_lo_addr_offset
+      end;
+      hi_addr_ptr = function (self)
+         return ffi.cast(uint8_ptr, self) + flow_key_ipv6_hi_addr_offset
+      end;
+   }
+})
+
+-- Helper functions
+
+local function ihl(p, offset)
+   local ver_and_ihl = p.data[offset]
+   return band(ver_and_ihl, 0x0F) * 4
+end
+
+
+--
+-- Traverse an IPv6 header which has the following layout:
+--
+--     0         8        16
+--     | NextHdr | HdrLen | ...
+--
+--  where "NextHdr" is the type code of the next header, and "HdrLen" is the
+--  length of the header in 8-octet units, sans the first 8 octets.
+--
+local function ipv6_nexthdr_type_len_skip (p)
+   return p[0], p + 8 + (p[1] * 8)
+end
+
+local ipv6_walk_header_funcs = {
+   [IPv6_NEXTHDR_HOPBYHOP] = ipv6_nexthdr_type_len_skip;
+   [IPv6_NEXTHDR_ROUTING]  = ipv6_nexthdr_type_len_skip;
+   [IPv6_NEXTHDR_DSTOPTS]  = ipv6_nexthdr_type_len_skip;
+   [IPv6_NEXTHDR_FRAGMENT] = function (p)
+      return p[0], p + 8
+   end;
+   [IPv6_NEXTHDR_AH] = function (p)
+      -- Size specified in 4-octet units (plus two octets).
+      return p[0], p + 2 + (p[1] * 4)
+   end;
+}
+
+--
+-- Traverses all the IPv6 headers (using the "next header" fields) until an
+-- upper-level protocol header (e.g. TCP, UDP) is found. The returned value
+-- is the type of the upper level protocol code and pointer to the beginning
+-- of the upper level protocol header data.
+--
+local function ipv6_walk_headers (p)
+   local ptr = p.data
+   local nexthdr = ptr[IPv6_NEXTHDR_OFFSET]
+   while ipv6_walk_header_funcs[nexthdr] do
+      local new_nexthdr, new_ptr = ipv6_walk_header_funcs[nexthdr](ptr)
+      if new_ptr > p.data + p.length then
+         break
+      end
+      nexthdr, ptr = new_nexthdr, new_ptr
+   end
+   return nexthdr, ptr
+end
+
+
+Scanner = subClass()
+Scanner._name = "SnabbWall base packet Scanner"
+
+function Scanner:extract_packet_info(p)
+   local eth_type  = rd16(p.data + ETH_TYPE_OFFSET)
+   local ip_offset = ETH_HEADER_SIZE
+   local vlan_id   = 0
+
+   while eth_type == ETH_TYPE_VLAN do
```

Comment:
Using a loop also manages to handle [VLAN double tagging](https://en.wikipedia.org/wiki/IEEE_802.1Q#Double_tagging). If a `while` loop looks like a bad idea, it could be changed it to a pair of `if` staments.